### PR TITLE
add migration from vintage secret format to org namespace format

### PR DESCRIPTION
**Pull request by anvddriesch**
Towards https://github.com/giantswarm/roadmap/issues/2395

The default dex config secret used to always have the same name and be located in each cluster namespace. However, with app secrets being located in org namespaces instead of cluster namespaces, this should be changed. If two clusters in the same organization have dex installed, they should still have a separate set of credentials.
This PR changes the default dex config secret name to include the app name (since it is unique within the namespace) to solve this.
Furthermore, it adds a migration code that copies existing configuration from the old to the new secret and gets rid of the old one.

### CODE REVIEW

File: pkg/idp/util.go [hunk 1]
```diff
@@ -69,7 +70,15 @@ func connectorsDefinedInUserConfigMap(userConfigmap *corev1.ConfigMap) bool {
 	return false
 }
 
-func GetDexSecretConfig(namespace string) v1alpha1.AppExtraConfig {
+func GetDexSecretConfig(n types.NamespacedName) v1alpha1.AppExtraConfig {
```
Comment (anvddriesch) [inline pkg/idp/util.go:73]: This is the new format

File: pkg/idp/util.go [hunk 2]
```diff
@@ -69,7 +70,15 @@ func connectorsDefinedInUserConfigMap(userConfigmap *corev1.ConfigMap) bool {
...
+	return v1alpha1.AppExtraConfig{
+		Kind:      "secret",
+		Name:      key.GetDexConfigName(n.Name),
+		Namespace: n.Namespace,
+		Priority:  25}
+}
+
+func GetVintageDexSecretConfig(namespace string) v1alpha1.AppExtraConfig {
```
Comment (anvddriesch) [inline pkg/idp/util.go:81]: Renaming the old one to vintage so it can be used in migration code

File: controllers/app_controller_test.go [hunk 3]
```diff
@@ -164,5 +163,106 @@ var _ = Describe("App controller", func() {
 			}, timeout, interval).Should(BeTrue())
 		})
 	})
+	const (
+		SecondAppNamespace = "test-namespace-2"
+	)
+
+	Context("When reconciling an app with vintage dex config secret", func() {
```
Comment (anvddriesch) [inline controllers/app_controller_test.go:170]: Adding a test for this

File: pkg/idp/idp.go [hunk 4]
```diff
@@ -107,11 +113,11 @@ func (s *Service) Reconcile(ctx context.Context) error {
 		if !apierrors.IsNotFound(err) {
 			return microerror.Mask(err)
 		} else {
-			secret = s.GetDefaultDexConfigSecret(dexSecretConfig.Name, dexSecretConfig.Namespace)
+			secret = GetDefaultDexConfigSecret(dexSecretConfig.Name, dexSecretConfig.Namespace)
 			if err := s.Create(ctx, secret); err != nil {
 				return microerror.Mask(err)
 			}
-			s.log.Info("Applied default dex config secret for dex app instance.")
+			s.log.Info("Created default dex config secret for dex app instance.")
```
Comment (anvddriesch) [inline pkg/idp/idp.go:120]: These logs where confusing because they said the same thing, so i changed them

File: pkg/idp/idp.go [hunk 5]
```diff
@@ -314,7 +320,7 @@ func (s *Service) GetAppConfig(ctx context.Context) (provider.AppConfig, error)
 	}, nil
 }
 
-func (s *Service) GetDefaultDexConfigSecret(name string, namespace string) *corev1.Secret {
+func GetDefaultDexConfigSecret(name string, namespace string) *corev1.Secret {
```
Comment (anvddriesch) [inline pkg/idp/idp.go:323]: There was no reason for this function to be a method and it made using it in tests more difficult

File: pkg/idp/idp.go [hunk 6]
```diff
@@ -353,3 +359,52 @@ func GetIssuerAddress(baseDomain string, managementClusterIssuerAddress string,
 	}
 	return issuerAddress
 }
+
+func (s *Service) migrateVintageSecret(ctx context.Context) error {
```
Comment (anvddriesch) [inline pkg/idp/idp.go:363]: migration code is here

**Review by mogottsch — APPROVED** [review]

**Pull request merged**